controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -243,41 +272,82 @@ private Job deploymentJob(Application application, JobType jobType, String reaso
                         Duration delay = ((DeploymentSpec.Delay) step).duration();
                         completedAt = completedAt.map(at -> at.plus(delay)).filter(at -> ! at.isAfter(clock.instant()));
                         reason += " after a delay of " + delay;
                     }
                     else {
-                        completedAt = stepJobs.stream().map(job -> completedAt(application.change(), application, job).get()).max(naturalOrder());
+                        completedAt = jobsByCompletion.keySet().stream().map(Optional::get).max(naturalOrder());
                         reason = "Available change in " + stepJobs.stream().map(JobType::jobName).collect(joining(", "));
                     }
                 }
-                else if (completedAt.isPresent()) { // Step not complete, because some jobs remain -- trigger these if the previous step was done.
-                    for (JobType job : remainingJobs)
-                        jobs.add(deploymentJob(application, job, reason, completedAt.get(), stepJobs));
-                    completedAt = Optional.empty();
-                    break;
-                }
             }
+
+            if (testJobs == null)
+                testJobs = testJobsFor(application, targetFor(application, application.change(), empty()), "Testing last changes outside prod", clock.instant());
+            jobs.addAll(testJobs);
+
             // TODO jvenstad: Replace with completion of individual parts of Change.
-            if (completedAt.isPresent())
+            if (steps.stream().flatMap(step -> step.zones().stream()).map(order::toJob)
+                     .allMatch(job -> completedAt(application.change(), application, job).isPresent()))
                 applications().lockIfPresent(id, lockedApplication -> applications().store(lockedApplication.withChange(Change.empty())));
         });
         return jobs;
     }
 
+    private List<Job> testJobsFor(Application application, State target, String reason, Instant availableSince) {
+        List<Step> steps = application.deploymentSpec().steps();
+        if (steps.isEmpty()) steps = singletonList(new DeploymentSpec.DeclaredZone(test));
+        List<Job> jobs = new ArrayList<>();
+        for (Step step : steps.stream().filter(step -> step.deploysTo(test) || step.deploysTo(staging)).collect(toList())) {
+            for (JobType jobType : step.zones().stream().map(order::toJob).collect(toList())) {
+                Optional<JobRun> completion = successOn(application, jobType, target);
+                if (completion.isPresent())
+                    availableSince = completion.get().at();
+                else if (isVerified(application, target, jobType))
+                    jobs.add(deploymentJob(application, target, application.change(), jobType, reason, availableSince, emptySet()));
+            }
+        }
+        return jobs;
+    }
+
+    private boolean isVerified(Application application, State state, JobType jobType) {
```

Comment:
I think `isTested` would be a better name.

Reply:
Sure.